File: apps/desktop/src/main/following.ts
// The socket that reads one running Check's log as it is written, held here
// beside `journal.ts` and for its reasons.
//
// **One at a time.** A person reads one Check's log; opening another replaces
// the first rather than holding two sockets against Fleet.
//
// **It reads and it cannot do anything else.** There is no send on this socket
// and no route behind it that takes one.

import WebSocket from "ws";

import type { FollowedLog, OutputMessage } from "@armada/protocol";
import { HOST } from "./runtime-file";

/**
 * How many lines main holds for the reader — the window a recorded Check log is
 * read in, so a live one and a recorded one show a person the same amount.
 */
const HELD = 2_000;

/** One running Check's log connection. */
export class FollowSocket {
  private readonly publish: (followed: FollowedLog) => void;
  private socket: WebSocket | null = null;
  private held: FollowedLog = { state: "none" };

  constructor(publish: (followed: FollowedLog) => void) {
    this.publish = publish;
  }

  /**
   * Which Check's log is being read, or `null` for either to stop.
   *
   * **The same log asked for again is the same reading.** A screen asks on
   * every render that names it, and a reading restarted each time would draw
   * its first window over and over.
   */
  open(port: number | null, jobId: string | null, kept: string | null): void {
    const held = this.held;
    if (
      jobId !== null &&
      kept !== null &&
      held.state !== "none" &&
      held.state !== "failed" &&
      held.jobId === jobId &&
      held.kept === kept
    ) {
      return;
    }
    this.close();
    if (jobId === null || kept === null) {
      this.set({ state: "none" });
      return;
    }
    if (port === null) {
      this.set({ state: "failed", jobId, kept, detail: "Fleet is not connected." });
      return;
    }
    this.set({ state: "opening", jobId, kept });

    const path = `/jobs/${encodeURIComponent(jobId)}/checks/${encodeURIComponent(kept)}/observe`;
    const socket = new WebSocket(`ws://${HOST}:${port}${path}`);
    this.socket = socket;
    socket.on("message", (data: WebSocket.RawData) => this.arrived(jobId, kept, String(data)));
    socket.on("error", (cause: Error) => this.broke(jobId, kept, cause.message));
    socket.on("close", () => this.broke(jobId, kept, "the connection closed"));
  }

  close(): void {
    const socket = this.socket;
    this.socket = null;
    if (socket === null) return;
    socket.removeAllListeners();
    // One listener stays, for `journal.ts`'s reason: a socket still shaking
    // hands reports its abort as an `error` on the next tick, and an `error`
    // with nothing listening is thrown by Node.
    socket.on("error", () => {});
    socket.close();
  }

  private set(followed: FollowedLog): void {
    this.held = followed;
    this.publish(followed);
  }

  private arrived(jobId: string, kept: string, text: string): void {
    let message: OutputMessage;
    try {
      message = JSON.parse(text) as OutputMessage;
    } catch {
      this.broke(jobId, kept, "Fleet sent a message this Bridge could not read.");
      return;
    }

    if (message.message === "opened") {
      this.set({
        state: "following",
        jobId,
        kept,
        name: message.name,
        attempt: message.attempt,
        path: message.path,
        fromLine: message.skipped + 1,
        lines: [],
      });
      return;
    }

    const held = this.held;
    if (held.state !== "following") return;

    if (message.message === "lines") {
      const all = [...held.lines, ...message.lines];
      const dropped = Math.max(0, all.length - HELD);
      this.set({ ...held, lines: all.slice(dropped), fromLine: held.fromLine + dropped });
      return;
    }

    // `closed` carries why, and the socket is let go first so its own `close`
    // cannot overwrite the reason — `journal.ts` learned this the hard way.
    this.close();
    this.set({ ...held, ended: message.because });
  }

  /**
   * The socket went without a sentence. **What had arrived stays arrived**, for
   * `journal.ts`'s reason: a reading that emptied on a dropped connection would
   * read as a Check that had printed nothing.
   */
  private broke(jobId: string, kept: string, detail: string): void {
    if (this.socket === null) return;
    this.socket.removeAllListeners();
    this.socket = null;
    const held = this.held;
    if (held.state === "following") {
      this.set({ ...held, ended: held.ended ?? "broke" });
      return;
    }
    this.set({ state: "failed", jobId, kept, detail });
  }
}

File: crates/api/src/following.rs
//! Watching one running Check's log: a reader Fleet implements, and the socket
//! that follows it.
//!
//! **`crate::journal`'s shape, one file over.** The runner appends a Check's
//! output to a file as it arrives, and this reads that file the way the Job's
//! log is read: what is already there, then what is appended, [`FOLLOW`]
//! behind. A request-response read of a growing file would need Bridge to ask
//! again on a timer, and Bridge does not poll.
//!
//! **Nothing on this stream can be missed**, for the journal's reason: the
//! cursor is a byte offset into the file, so a viewer that stalls reads the gap
//! on its next pass. There is no queue between the runner and the socket.
//!
//! **`api` does not open the file.** [`Follow`] is stated here and implemented
//! in `fleet`, which knows where a live log is and when the Check writing it
//! has ended. What crosses is `ipc` vocabulary.

use std::sync::Arc;

use axum::extract::ws::{Message, WebSocket};
use ipc::{
    JobId, OutputClosed, OutputEnded, OutputLines, OutputMessage, OutputOpened, PROTOCOL_VERSION,
};

use crate::journal::FOLLOW;

/// One running Check's log, read. **Implemented in Fleet.**
///
/// Blocking, for [`Journal`](crate::Journal)'s reason: each method is a file
/// read or a lock, and the socket calls them off the runtime's workers.
pub trait Follow: Send + Sync + 'static {
    /// Everything written after `from`, and where the next pass starts.
    ///
    /// **Whole lines only, unless `to_the_end`.** A line still being written
    /// stays in the file until its newline arrives, so a reader never draws half
    /// a line the next pass would have to take back. The last pass, once the
    /// Check has ended, takes whatever is left.
    ///
    /// **The first pass is bounded**, and `skipped` says how many lines older
    /// than the window it left out.
    fn read(&self, from: u64, to_the_end: bool) -> Followed;

    /// Whether the Check is still writing. **`false` is final**: the Check has
    /// ended, and the next pass that reads to the end is the last one.
    fn writing(&self) -> bool;
}

/// What one pass over a live log came to.
pub struct Followed {
    /// The lines, oldest first, verbatim.
    pub lines: Vec<String>,
    /// The byte the next pass starts at. Never past a partial line unless the
    /// pass was asked to read to the end.
    pub from: u64,
    /// Lines older than the first pass's window, left out of it.
    pub skipped: u64,
    /// The file is there and could not be read. **Not a file not yet made** —
    /// a Check that has just started may not have opened it, and that answers
    /// with nothing.
    pub unreadable: bool,
}

/// A running Check's log, resolved: whose it is, and the reader for it.
///
/// **What the daemon answers before the socket opens**, for `observe_job`'s
/// reason: a name that resolves to nothing is a refusal the caller reads at the
/// moment it asked rather than a socket that opens and says nothing.
#[derive(Clone)]
pub struct LiveOutput {
    /// The Check whose log this is.
    pub name: String,
    /// Which run of the step is writing it.
    pub attempt: u32,
    /// Where it is being written, relative to the repository root.
    pub path: String,
    pub follow: Arc<dyn Follow>,
}

/// Serve one viewer: what the log holds, then what is appended, then why it
/// stopped.
///
/// **Whether the Check is still writing is asked before each pass rather than
/// after**, so the pass that follows a `false` reads to the end — a Check that
/// finished between two passes still has its last line sent.
pub(crate) async fn relay(mut socket: WebSocket, job_id: JobId, live: LiveOutput) {
    let mut writing = still(&live.follow).await;
    let first = pass(&live.follow, 0, !writing).await;
    let opened = OutputOpened {
        protocol_version: PROTOCOL_VERSION,
        job_id,
        name: live.name.clone(),
        attempt: live.attempt,
        path: live.path.clone(),
        skipped: first.skipped,
    };
    if !send(&mut socket, &OutputMessage::Opened(opened)).await {
        return;
    }
    let mut from = first.from;
    let mut reading = first;
    loop {
        if reading.unreadable {
            closed(&mut socket, OutputEnded::Unreadable).await;
            return;
        }
        if !reading.lines.is_empty() {
            let lines = OutputLines {
                lines: reading.lines,
            };
            if !send(&mut socket, &OutputMessage::Lines(lines)).await {
                return;
            }
        }
        if !writing {
            closed(&mut socket, OutputEnded::Finished).await;
            return;
        }
        tokio::time::sleep(FOLLOW).await;
        writing = still(&live.follow).await;
        reading = pass(&live.follow, from, !writing).await;
        from = reading.from;
    }
}

async fn still(follow: &Arc<dyn Follow>) -> bool {
    let follow = Arc::clone(follow);
    tokio::task::spawn_blocking(move || follow.writing())
        .await
        .unwrap_or(false)
}

/// One pass, off the runtime's worker threads, for the journal's reason.
async fn pass(follow: &Arc<dyn Follow>, from: u64, to_the_end: bool) -> Followed {
    let follow = Arc::clone(follow);
    match tokio::task::spawn_blocking(move || follow.read(from, to_the_end)).await {
        Ok(read) => read,
        // A panic in the reader, not anything about the log. Ending with a
        // reason is the honest answer; a viewer left waiting would read it as a
        // Check that had gone quiet.
        Err(_) => Followed {
            lines: Vec::new(),
            from,
            skipped: 0,
            unreadable: true,
        },
    }
}

async fn closed(socket: &mut WebSocket, because: OutputEnded) {
    send(socket, &OutputMessage::Closed(OutputClosed { because })).await;
}

async fn send(socket: &mut WebSocket, message: &OutputMessage) -> bool {
    let Ok(text) = ipc::encode(message) else {
        return false;
    };
    // Awaited, not queued, for the journal's reason.
    socket.send(Message::Text(text)).await.is_ok()
}
